src/Functions/vectorFunctions.cpp

```diff
@@ -0,0 +1,1067 @@
+
+        if (!cur_tuple)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Argument 0 of function {} should be tuples, got {}",
+                            getName(), arguments[0].type->getName());
+
+        const auto & cur_types = cur_tuple->getElements();
+
+        Columns cur_elements;
+        if (arguments[0].column)
+            cur_elements = getTupleElements(*arguments[0].column);
+
+        size_t tuple_size = cur_types.size();
+        if (tuple_size == 0)
+            return std::make_shared<DataTypeUInt8>();
+
+        const auto & p_column = arguments[1];
+        auto abs = FunctionFactory::instance().get("abs", context);
+        auto pow = FunctionFactory::instance().get("pow", context);
+        auto plus = FunctionFactory::instance().get("plus", context);
+        DataTypePtr res_type;
+        for (size_t i = 0; i < tuple_size; ++i)
+        {
+            try
+            {
+                ColumnWithTypeAndName cur{cur_elements.empty() ? nullptr : cur_elements[i], cur_types[i], {}};
+                auto elem_abs = abs->build(ColumnsWithTypeAndName{cur});
+                cur.type = elem_abs->getResultType();
+                cur.column = cur.type->createColumn();
+
+                auto elem_pow = pow->build(ColumnsWithTypeAndName{cur, p_column});
+
+                if (i == 0)
+                {
+                    res_type = elem_pow->getResultType();
+                    continue;
+                }
+
+                ColumnWithTypeAndName left_type{res_type, {}};
+                ColumnWithTypeAndName right_type{elem_pow->getResultType(), {}};
+                auto plus_elem = plus->build({left_type, right_type});
+                res_type = plus_elem->getResultType();
+            }
+            catch (DB::Exception & e)
+            {
+                e.addMessage("While executing function {} for tuple element {}", getName(), i);
+                throw;
+            }
+        }
+
+        auto divide = FunctionFactory::instance().get("divide", context);
+        ColumnWithTypeAndName one{std::make_shared<DataTypeUInt8>(), {}};
+        auto div_elem = divide->build({one, p_column});
+        ColumnWithTypeAndName inv_p_column{div_elem->getResultType(), {}};
+        return pow->build({ColumnWithTypeAndName{res_type, {}}, inv_p_column})->getResultType();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        const auto * cur_tuple = checkAndGetDataType<DataTypeTuple>(arguments[0].type.get());
+        const auto & cur_types = cur_tuple->getElements();
+        auto cur_elements = getTupleElements(*arguments[0].column);
+
+        size_t tuple_size = cur_elements.size();
+        if (tuple_size == 0)
+            return DataTypeUInt8().createColumnConstWithDefaultValue(input_rows_count);
+
+        const auto & p_column = arguments[1];
```

Comment:
Is there use-cases for non-constnat `p`, how do you think? Also, shold we limit possible values for `p`, like `0 < p < inf` or make it integer? 

```
SELECT LpNorm((3, 1, 4), 0), LpNorm((3, 1, 4), inf);

┌─LpNorm((3, 1, 4), 0)─┬─LpNorm((3, 1, 4), inf)─┐
│                  inf │                      1 │
└──────────────────────┴────────────────────────┘
```

Reply:
I decided to not limit `p`. It definitely should NOT be just integer (I will add a test when `p` is float). Real math sense is when `p >= 1`, but probably somebody will want to use it for other purposes...

Reply:
Let's make it `1 <= p < inf`?

Reply:
Ok.

Reply:
Restrictions added in the 53d7649c33bb43410b4e1ad52b820da4eebb327f.